tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -0,0 +1,381 @@
+
+}
+
+
+#set up guest test environment
+sub prepare_guest() {
+  my $vm = shift;
+
+  assert_script_run "virsh dumpxml $vm > $vm.xml";
+  script_run "virsh destroy $vm";
+  assert_script_run "virsh undefine $vm";
+
+  #extra process for XEN hypervisor
+  if (get_var("XEN") || check_var("SYSTEM_ROLE", "xen") || check_var("HOST_HYPERVISOR", "xen")) {
+
+    #enable pci-passthrough and set model for pv guest.
+    #refer to bug #1167217 for the reason
+    my $passthru_xml = "<passthrough state='on'/>";
+    my $e820_xml = "";
+    if ($vm =~ /\bpv\b/) {
+      $e820_xml = "<e820_host state='on'/>";
+    }
+    if (script_run("grep '<features>' $vm.xml") == 0) {
+      assert_script_run "sed -i \"/<features>/a\\<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\" $vm.xml";
+    }
+    else {
+      assert_script_run "sed -i \"/<domain /a\\<features>\\n<xen>\\n$passthru_xml\\n$e820_xml\\n</xen>\\n</features>\" $vm.xml";
+    }
+
+    #disable memory ballooning for fv guest as it is not supported
+    if ($vm =~ /\bfv\b/ || $vm =~ /\bhvm\b/) {
+      assert_script_run "sed -i '/<currentMemory/d' $vm.xml";
+    }
+   
+  }
+
+  #add pcie controllers to support hotplugging more SR-IOV Ethernet vf devices
+  my $controller_xml = "<controller type='pci' model='pcie-root-port'/>";
+  assert_script_run "sed -i \"/<devices>/a\\$controller_xml\\n$controller_xml\\n$controller_xml\" $vm.xml";
+  assert_script_run "virsh define $vm.xml";
+  assert_script_run "virsh start $vm";
+  script_run "virsh dumpxml $vm > $vm.xml";
+  upload_logs("$vm.xml", log_name => "initial");
+
+  #passwordless access to guest.
+  script_run "sed -i '/192.168.123/d' /root/.ssh/known_hosts";    #just for julie's tests, should remove before submit
+#  script_run "ssh-keygen -R sles-15-sp2-64-fv-def-net -f /root/.ssh/known_hosts";
+  save_guest_ip($vm, name => "br123");     #get the guest ip via key words in 'virsh domiflist'
+  exec_and_insert_password("ssh-copy-id -o StrictHostKeyChecking=no -i ~/.ssh/id_rsa.pub -f root\@$vm");
+
+  #set default interface for guest (or hvm guest on XEN may reboot with interfaces randomly)
+  #seems need to fix the pci slot and put it as top as possible
+  #my $mac_vm = script_output("virsh domiflist $vm | grep "br123" | grep -oE \"[[:xdigit:]]{2}(:[[:xdigit:]]{2}){5}\"");
+  #my $nic_vm = script_output "ssh root\@$vm \"grep '$mac_vm' /sys/class/net/*/address | cut -d'/' -f5 | head -n1\""
+  #assert_script_run("ssh root\@$vm \"echo \\\"BOOTPROTO=\\'dhcp\\'\\nSTARTMODE=\\'auto\\'\\nLLADDR=\\'$mac_vm\\'\\\" > /etc/sysconfig/network/ifcfg-$nic_vm\"");
+
+  #install required packages on guest
+  #script_output "ssh root\@$vm \"zypper -t in pciutils\"";  #julie: note to enable it finally, but found redundante?
+}
+
+
+#detach a specified vf Ethernet device from host
+sub detach_vf_from_host(){
+  my $device_bdf = shift;
+
+  #change to device id in libvirt
+  $device_bdf =~ s/[:\.]/_/g;
+  my $device_id = script_output "virsh nodedev-list | grep $device_bdf";
```

Comment:
Does this grep $device_bdf work ?

Reply:
what do you mean? you suspect this shell command does not work, or you think just 'grep $device_bdf' is enough? 
virsh nodedev-list will list all the pci devices, then grep get the specified vf device.